File: spyweb/solver/replay.py
from dataclasses import dataclass

import numpy as np

from spyweb.core.events import (
    AccusationResolved,
    QuestionAnswered,
    SecondAnswerBought,
    Trace,
    TraceStep,
)
from spyweb.solver.belief import Belief, filter_first_answer, filter_paid_second, pair_count
from spyweb.solver.encoding import Encoding
from spyweb.solver.universe import Universe


@dataclass(frozen=True)
class ReplayState:
    belief: Belief
    trace: Trace = ()


def apply_event(
    universe: Universe, encoding: Encoding, state: ReplayState, event: object
) -> ReplayState:
    before = state.belief
    after = before
    if isinstance(event, QuestionAnswered):
        after = filter_first_answer(
            universe,
            before,
            encoding.question_id(event.question),
            encoding.answer_code(event.answer),
        )
    elif isinstance(event, SecondAnswerBought):
        after = filter_paid_second(
            universe,
            before,
            encoding.question_id(event.question),
            encoding.answer_code(event.first),
            encoding.answer_code(event.second),
        )
    elif isinstance(event, AccusationResolved):
        matches = (universe.ringleader[before] == event.ringleader) & (
            universe.hideout[before] == event.hideout
        )
        after = before[matches if event.correct else np.logical_not(matches)]
    else:
        raise TypeError("Unsupported observed event")
    step = TraceStep(
        len(state.trace) + 1,
        event,
        int(before.size),
        int(after.size),
        pair_count(universe, before),
        pair_count(universe, after),
    )
    return ReplayState(after, (*state.trace, step))

File: src/solver/replay.ts
import type { GameEvent, TraceStep } from "../core/events.js";
import type { Encoding } from "./encoding.js";
import {
  filterFirstAnswer,
  filterPaidSecondAnswer,
  pairCount,
  type Belief
} from "./belief.js";
import type { CompactUniverse } from "./universe.js";

export interface ReplayState {
  readonly belief: Belief;
  readonly trace: readonly TraceStep[];
}

export function applyObservedEvent(
  universe: CompactUniverse,
  encoding: Encoding,
  state: ReplayState,
  event: GameEvent
): ReplayState {
  const before = state.belief;
  let after = before;
  if (event.type === "question-answered") {
    after = filterFirstAnswer(
      universe,
      before,
      encoding.questionId(event.question),
      encoding.answerCode(event.answer)
    );
  } else if (event.type === "second-answer-bought") {
    after = filterPaidSecondAnswer(
      universe,
      before,
      encoding.questionId(event.question),
      encoding.answerCode(event.first),
      encoding.answerCode(event.second)
    );
  } else if (event.type === "accusation-resolved") {
    const matches: number[] = [];
    for (const board of before) {
      const isPair =
        universe.ringleader[board] === event.ringleader &&
        universe.hideout[board] === event.hideout;
      if (isPair === event.correct) matches.push(board);
    }
    after = Uint32Array.from(matches);
  }
  const step: TraceStep = {
    sequence: state.trace.length + 1,
    event,
    boardsBefore: before.length,
    boardsAfter: after.length,
    pairsBefore: pairCount(universe, before),
    pairsAfter: pairCount(universe, after)
  };
  return { belief: after, trace: [...state.trace, step] };
}
